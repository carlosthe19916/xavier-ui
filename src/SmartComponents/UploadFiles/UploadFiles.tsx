import './UploadFiles.scss';
import React, { Component, createRef } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { withRouter, Link } from 'react-router-dom';
import {
    FolderOpenIcon
} from '@patternfly/react-icons';
import {
    Card,
    CardBody,
    Button,
    CardFooter,
    Progress,
    ProgressVariant,
    Stack,
    StackItem,
    FormGroup,
    TextInput,
    Form
} from '@patternfly/react-core';

import Dropzone from 'react-dropzone';
import UploadFilesPage from '../../PresentationalComponents/UploadFilesPage/UploadFilesPage';
import {
    uploadRequest,
    uploadProgress,
    uploadClear
} from '../../actions/UploadActions';
import { RouterGlobalProps } from '../../models/router';
import { GlobalState } from '../../models/state';
import { Upload } from '../../models/Upload';

interface Props extends RouterGlobalProps {
    uploadProgress: (file: File, progress: number) => void;
    uploadRequest: (customerId: string, file: File, config: {}) => void;
    uploadClear: () => void;
    uploads: Upload[];
};

interface State {
    customerId: string;
};

export class UploadFiles extends Component<Props, State> {

    constructor(props: Props) {
        super(props);
        this.onDrop = this.onDrop.bind(this);
        this.clearUploads();

        // TODO replace this to start using the customerId within the session
        this.state = {
            customerId: '123456'
        };
    }

    onDrop(files: File[]): void {
        this.startUpload(files);
    };

    startUpload(files: File[]): void {
        files.forEach((file) => {
            const config = {
                headers: {
                    'X-Requested-With': 'XMLHttpRequest'
                },
                onUploadProgress: (progressEvent: any) => {
                    const progress: number = Math.round((progressEvent.loaded / progressEvent.total) * 100);
                    this.props.uploadProgress(file, progress);
                }
            };

            const customerId: string = this.state.customerId;
            this.props.uploadRequest(customerId, file, config);
        });
    }

    clearUploads(): void {
        this.props.uploadClear();
    }

    render() {
        const showDropZone: boolean = this.props.uploads.length === 0;
        const uploading: boolean = this.props.uploads.some((element) => element.uploading);

        const dropzoneRef: any = createRef();
        const openDialog = () => {
            if (dropzoneRef.current) {
                dropzoneRef.current.open();
            }
        };

        return (
            <UploadFilesPage
                title='Upload'
                showBreadcrumb={ false }>
                <Card>
                    <CardBody>
                        {
                            showDropZone ?
                                <div>
                                    <Dropzone
                                        onDrop={ this.onDrop }
                                        ref={ dropzoneRef }
                                        noClick noKeyboard
                                        multiple={ false }
                                        accept={ [ 'application/zip', 'application/json' ] }
                                    >
                                        { ({ getRootProps, getInputProps }) => {
                                            return (
                                                <div className="container">
                                                    <div { ...getRootProps({ className: 'dropzone' }) }>
                                                        <input { ...getInputProps() } />
                                                        <FolderOpenIcon className="pf-c-title pf-m-4xl" />
                                                        <p className="pf-c-title pf-m-3xl">Drag a file here</p>
                                                        <p className="pf-c-title">- or -</p>
                                                        <Button
                                                            type="button"
                                                            variant="tertiary"
                                                            onClick={ openDialog }>
                                                            Select a file from your computer
                                                        </Button>
                                                    </div>
                                                </div>
                                            );
                                        } }
                                    </Dropzone>
                                    <br/>
                                    <Form isHorizontal className="pf-l-level">
                                        <FormGroup
                                            label="Customer id"
                                            isRequired
                                            fieldId="horizontal-form-name"
                                        >
                                            <TextInput
                                                value={ this.state.customerId }
                                                isRequired
                                                type="text"
                                                id="horizontal-form-name"
                                                aria-describedby="horizontal-form-name-helper"
                                                name="horizontal-form-name"
                                            />
                                        </FormGroup>
                                    </Form>
                                </div>
                                :
                                <Stack gutter="md">
                                    {
                                        this.props.uploads.map((upload) => {
                                            return (
                                                <StackItem isMain key={ upload.file.name }>
                                                    <Progress
                                                        value={ upload.progress }
                                                        title={ upload.error
                                                            ? `${upload.file.name}: ${upload.error}`
                                                            : `Uploading: ${upload.file.name}` }
                                                        variant={ upload.error
                                                            ? ProgressVariant.danger
                                                            : ProgressVariant.info } />
                                                </StackItem>
                                            );
                                        })
                                    }
                                </Stack>
                        }
                    </CardBody>
                    <CardFooter>
                        { (showDropZone || uploading) ?
                            <Button variant="tertiary" component={ Link } to="/reports">Cancel</Button> :
                            <Button variant="primary" component={ Link } to="/reports">Next</Button>
                        }
                    </CardFooter>
                </Card>
            </UploadFilesPage>
        );
    }
}

const mapStateToProps = (state: GlobalState)  => {
    let { uploads: { uploads }} = state;
    return {
        uploads
    };
};

const mapDispatchToProps = (dispatch: any) =>
    bindActionCreators({
        uploadRequest,
        uploadProgress,
        uploadClear
    }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(UploadFiles));
