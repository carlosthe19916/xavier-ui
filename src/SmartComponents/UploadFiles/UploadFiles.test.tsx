import { shallow } from "enzyme";
import { UploadFiles } from "./UploadFiles";

const props = {
  history: null,
  location: null,
  match: null,

  uploadProgress: jest.fn(),
  uploadRequest: jest.fn(),
  uploadClear: jest.fn(),
  uploads: [
    {
      file: new File([""], "myFile1.zip"),
      success: null,
      error: null,
      progress: null,
      uploading: false
    },
    {
      file: new File([""], "myFile2.zip"),
      success: null,
      error: null,
      progress: null,
      uploading: false
    }
  ]
};

describe("UploadFiles", () => {
  it("expect to render Dropzone", () => {
    const emptyProps = Object.assign({}, props, {
      uploads: []
    });
    const wrapper = shallow(<UploadFiles {...emptyProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("expect to render progress bar", () => {
    const emptyProps = Object.assign({}, props, {
      uploads: [
        ...props.uploads,
        {
          file: new File([""], "myFile3.zip"),
          success: null,
          error: null,
          progress: 30,
          uploading: true
        }
      ]
    });
    const wrapper = shallow(<UploadFiles {...emptyProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  // it("expect to redirect because error", () => {
  //   const errorProps = Object.assign({}, props, {
  //     error: 'Error loading data'
  //   });
  //   const wrapper = shallow(<UploadFiles {...errorProps} />);
  //   expect(wrapper).toMatchSnapshot();
  // });

  // it("expect to render loading object", () => {
  //   const loadingProps = Object.assign({}, props, {
  //     loading: true
  //   });
  //   const wrapper = shallow(<UploadFiles {...loadingProps} />);
  //   expect(wrapper).toMatchSnapshot();
  // });

  // it("expect to render object", () => {
  //   const objProps = Object.assign({}, props);
  //   const wrapper = shallow(<UploadFiles {...objProps} />);
  //   expect(wrapper).toMatchSnapshot();
  // });
});
