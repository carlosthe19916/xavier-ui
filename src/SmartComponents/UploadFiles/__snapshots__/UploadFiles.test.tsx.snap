// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadFiles expect to render Dropzone 1`] = `
<withRouter(UploadFilesPage)
  showBreadcrumb={false}
  title="Upload"
>
  <Card
    className=""
    component="article"
  >
    <CardBody
      className=""
      component="div"
    >
      <div>
        <Dropzone
          accept={
            Array [
              "application/zip",
              "application/json",
            ]
          }
          multiple={false}
          noClick={true}
          noKeyboard={true}
          onDrop={[Function]}
        >
          <Component />
        </Dropzone>
        <br />
        <Form
          className="pf-l-level"
          isHorizontal={true}
        >
          <FormGroup
            className=""
            fieldId="horizontal-form-name"
            isInline={false}
            isRequired={true}
            isValid={true}
            label="Customer id"
          >
            <TextInput
              aria-describedby="horizontal-form-name-helper"
              aria-label={null}
              className=""
              id="horizontal-form-name"
              isDisabled={false}
              isReadOnly={false}
              isRequired={true}
              isValid={true}
              name="horizontal-form-name"
              onChange={[Function]}
              type="text"
              value="123456"
            />
          </FormGroup>
        </Form>
      </div>
    </CardBody>
    <CardFooter
      className=""
      component="div"
    >
      <Button
        aria-label={null}
        className=""
        component={[Function]}
        isActive={false}
        isBlock={false}
        isDisabled={false}
        isFocus={false}
        isHover={false}
        to="/reports"
        type="button"
        variant="tertiary"
      >
        Cancel
      </Button>
    </CardFooter>
  </Card>
</withRouter(UploadFilesPage)>
`;

exports[`UploadFiles expect to render progress bar 1`] = `
<withRouter(UploadFilesPage)
  showBreadcrumb={false}
  title="Upload"
>
  <Card
    className=""
    component="article"
  >
    <CardBody
      className=""
      component="div"
    >
      <Stack
        className=""
        component="div"
        gutter="md"
      >
        <StackItem
          className=""
          isMain={true}
          key="myFile1.zip"
        >
          <Progress
            className=""
            id=""
            max={100}
            measureLocation="top"
            min={0}
            size={null}
            title="Uploading: myFile1.zip"
            value={null}
            valueText={null}
            variant="info"
          />
        </StackItem>
        <StackItem
          className=""
          isMain={true}
          key="myFile2.zip"
        >
          <Progress
            className=""
            id=""
            max={100}
            measureLocation="top"
            min={0}
            size={null}
            title="Uploading: myFile2.zip"
            value={null}
            valueText={null}
            variant="info"
          />
        </StackItem>
        <StackItem
          className=""
          isMain={true}
          key="myFile3.zip"
        >
          <Progress
            className=""
            id=""
            max={100}
            measureLocation="top"
            min={0}
            size={null}
            title="Uploading: myFile3.zip"
            value={30}
            valueText={null}
            variant="info"
          />
        </StackItem>
      </Stack>
    </CardBody>
    <CardFooter
      className=""
      component="div"
    >
      <Button
        aria-label={null}
        className=""
        component={[Function]}
        isActive={false}
        isBlock={false}
        isDisabled={false}
        isFocus={false}
        isHover={false}
        to="/reports"
        type="button"
        variant="tertiary"
      >
        Cancel
      </Button>
    </CardFooter>
  </Card>
</withRouter(UploadFilesPage)>
`;
